### ISSUE TITLE
Document the UrlResolver interface

### ISSUE THREAD

**Issue by jbdeboer** [angular/angular#1732]
[ <3 GT, but not in the critical path ]

The ```UrlResolver``` API is not documented and has non-obvious behaviour with respect to ```null``` inputs.

The interface should be documented and explain when ```baseUrl``` and ```url``` can be null and what that actually means.

**Comment by matanlurey** [comment]
Ping on this. Is it possible to just a quick explanation (in this issue) of what the two are? Thanks!

**Comment by vicb** [comment]
When the `baseUrl` is null, the `url` is resolved in the context of the current document:
- `./path` would resolve to `path/to/doc/path`,
- `http://custom/url` would resolve to `http://custom/url`

When the `url` is null and `baseUrl` specified, it resolves to `baseUrl`

Note that is is not possible to resolve a `baseUrl` + an `url` starting with `/` and the reason is that when you have:
- `baseUrl` = `http://www.foo.com/path/to/app`
- `url` = `/custom/path`
it would not play well with the url resolver because it uses a `<a>` for resolving URLs. It used to work in my first implementation before decided to switch to resolving via a link with @mhevery.

If this helps, I'll add the information as inline docs. Let me know any further details you would need.

**Comment by matanlurey** [comment]
Thanks, that is a good start.

**Comment by vicb** [comment]
Submitted a PR could one of you review it ? Thanks.

**Comment by jbdeboer** [comment]
Your explanation is good, except it is describing the default implementation in Angular. We are confused because we need to override that implementation.

A specific question that is not answered by #1863: In what cases will ```baseUrl``` be null?

**Comment by vicb** [comment]
From the PR:

> When `baseUrl` is null `url` is resolved in the context of the current document.

IIRC this is used only once to get the application URL:

We call [`this._appUrl = urlResolver.resolve(null, './');`](https://github.com/angular/angular/blob/0f002a5b18120d1f7bdfc7dca78f4e1c53794659/modules/angular2/src/core/compiler/compiler.js#L74) to get the base URL of the application

**Comment by jbdeboer** [comment]
If that is the case, it feels like the UrlResolver class is doing too many things and should be split into different interfaces.

**Comment by vicb** [comment]
@jbdeboer you're right. This is currently possible because the current implementation uses an anchor but  it would be much cleaner to have different interfaces.

**Comment by jbdeboer** [comment]
To summarize the current state of this bug:

@vicb gave an explanation for when ```baseUrl``` will be null: Only once to get the application URL.  Adding that information to the docs is enough to resolve this issue.

However, both @vicb and myself agree that this interface is doing too many things and could be refactored. That work could be tracked in a different issue and it is not a priority for GT.
